gcloud/monitoring/query.py

```diff
@@ -0,0 +1,661 @@
+        :param headers_only:
+             Whether to omit the point data from the time series objects.
+
+        :type _page_size: integer or None
+        :param _page_size:
+            An optional positive number specifying the maximum number of
+            points to return per page. This can be used to control how far
+            the iterator reads ahead.
+
+        :rtype: iterator over :class:`~gcloud.monitoring.timeseries.TimeSeries`
+        :returns: Time series objects, containing points ordered from oldest
+            to newest.
+        :raises: :exc:`ValueError` if the query time interval has not been
+            specified.
+        """
+        if self._end_time is None:
+            raise ValueError('Query time interval not specified.')
+
+        path = '/projects/{project}/timeSeries/'.format(
+            project=self._client.project)
+
+        def _fragments():
+            page_token = None
+            while True:
+                params = self._build_query_params(
+                    headers_only=headers_only,
+                    page_size=_page_size,
+                    page_token=page_token,
+                )
+                response = self._client.connection.api_request(
+                    method='GET',
+                    path=path,
+                    query_params=params,
+                )
+                for info in response.get('timeSeries', []):
+                    yield TimeSeries._from_dict(info)
+
+                page_token = response.get('nextPageToken')
+                if not page_token:
+                    break
+
+        # The following use of groupby() relies on equality comparison
+        # of time series as (named) tuples.
+        for timeseries, fragments in itertools.groupby(
+                _fragments(),
+                lambda fragment: fragment.header()):
+            points = list(itertools.chain.from_iterable(
+                fragment.points for fragment in fragments))
+            points.reverse()  # Order from oldest to newest.
+            yield timeseries.header(points=points)
+
+    def _build_query_params(self, headers_only=False,
+                            page_size=None, page_token=None):
+        """Assemble the list of key-value pairs for the URL query string.
+
+        We use a list of key-value pairs instead of a ``dict`` to allow for
+        repeated fields.
+
+        :type headers_only: boolean
+        :param headers_only:
+             Whether to omit the point data from the
+             :class:`~gcloud.monitoring.timeseries.TimeSeries` objects.
+
+        :type page_size: integer or None
+        :param page_size: A limit on the number of points to return per page.
+
+        :type page_token: string or None
+        :param page_token: A token to continue the retrieval.
+
+        :rtype: list of tuples
+        :returns:
+            A list of key-value pairs suitable for passing to ``urlencode``.
+        """
+        def _pairs():
```

Comment:
You could just have `_build_query_params` **be** the generator and force the callers to cast to a list if they need to.